modules/transport-netty4/src/main/java/org/elasticsearch/transport/NettyAllocator.java

```diff
@@ -28,32 +28,74 @@
+import org.elasticsearch.common.unit.ByteSizeValue;
 import org.elasticsearch.monitor.jvm.JvmInfo;
 
 public class NettyAllocator {
 
     private static final ByteBufAllocator ALLOCATOR;
+    private static final String DESCRIPTION;
 
     private static final String USE_UNPOOLED = "es.use_unpooled_allocator";
     private static final String USE_NETTY_DEFAULT = "es.unsafe.use_netty_default_allocator";
+    private static final String USE_NETTY_DEFAULT_CHUNK = "es.unsafe.use_netty_default_chunk_and_page_size";
 
     static {
         if (Booleans.parseBoolean(System.getProperty(USE_NETTY_DEFAULT), false)) {
             ALLOCATOR = ByteBufAllocator.DEFAULT;
+            DESCRIPTION = "[name=netty_default, factors={es.unsafe.use_netty_default_allocator=true}]";
         } else {
+            final long heapSizeInBytes = JvmInfo.jvmInfo().getMem().getHeapMax().getBytes();
+            final boolean g1gcEnabled = Boolean.parseBoolean(JvmInfo.jvmInfo().useG1GC());
+            final long g1gcRegionSizeInBytes = JvmInfo.jvmInfo().getG1RegionSize();
+            final boolean g1gcRegionSizeIsKnown = g1gcRegionSizeInBytes != -1;
+            ByteSizeValue heapSize = new ByteSizeValue(heapSizeInBytes);
+            ByteSizeValue g1gcRegionSize = new ByteSizeValue(g1gcRegionSizeInBytes);
+
             ByteBufAllocator delegate;
-            if (useUnpooled()) {
+            if (useUnpooled(heapSizeInBytes, g1gcEnabled, g1gcRegionSizeIsKnown, g1gcRegionSizeInBytes)) {
                 delegate = UnpooledByteBufAllocator.DEFAULT;
+                DESCRIPTION = "[name=unpooled, factors={es.unsafe.use_unpooled_allocator=" + userForcedUnpooled()
+                    + ", g1gc_enabled=" + g1gcEnabled
+                    + ", g1gc_region_size=" + g1gcRegionSize
+                    + ", heap_size=" + heapSize + "}]";
             } else {
                 int nHeapArena = PooledByteBufAllocator.defaultNumHeapArena();
-                int pageSize = PooledByteBufAllocator.defaultPageSize();
-                int maxOrder = PooledByteBufAllocator.defaultMaxOrder();
+                int pageSize;
+                int maxOrder;
+                if (useDefaultChunkAndPageSize()) {
+                    pageSize = PooledByteBufAllocator.defaultPageSize();
+                    maxOrder = PooledByteBufAllocator.defaultMaxOrder();
+                } else {
+                    pageSize = 8192;
+                    if (g1gcEnabled == false) {
+                        // This combined with a 8192 page size = 1 MB chunk sizes
+                        maxOrder = 7;
+                    } else if (g1gcRegionSizeIsKnown == false) {
+                        // This combined with a 8192 page size = 1 MB chunk sizes
+                        maxOrder = 7;
+                    } else if (g1gcRegionSizeInBytes >= (4 * 1024 * 1024)) {
+                        // This combined with a 8192 page size = 1 MB chunk sizes
+                        maxOrder = 7;
```

Comment:
nit: I would prefer to have these in a single if with `||` to avoid the repetition.